File: problog/library/string.pl
:- module(string, [sub_string/5, concat/2, join/3, str2lst/2, lst2str/2]).

:- use_module(library('string.py')).
:- use_module(library(lists)).

sub_string(String, Before, Length, After, SubString) :-
    str2lst(String, List),
    sub_list(List, BeforeL, Length, AfterL, SubListL),
    join('', BeforeL, Before),
    join('', AfterL, After),
    join('', SubListL, SubString).



    

File: problog/library/string.py
from problog.extern import problog_export
from problog.logic import unquote, make_safe, Term


@problog_export("+list", "-str")
def concat(terms):
    return make_safe("".join(map(lambda x: unquote(str(x)), terms)))


@problog_export("+str", "+list", "-str")
def join(sep, terms):
    return make_safe(unquote(sep).join(map(lambda x: unquote(str(x)), terms)))


@problog_export("+str", "-list")
def str2lst(string):
    return list(map(Term, map(make_safe, string)))


@problog_export("+list", "-str")
def lst2str(terms):
    return join("", terms)
